byterun/memory.c

```diff
@@ -686,29 +688,255 @@ CAMLexport CAMLweakdef void caml_modify (value *fp, value val)
+    uintnat *p3 = (void *) ((char *) *b + sz + Page_size);
+    for (p = p0; p < p1; p++) *p = Debug_filler_align;
+    for (p = p1; p < p2; p++) *p = Debug_uninit_align;
+    for (p = p2; p < p3; p++) *p = Debug_filler_align;
+  }
+#endif
+  return (char *) (aligned_mem - modulo);
+}
+
+/* [sz] and [modulo] are numbers of bytes */
+CAMLexport void* caml_stat_alloc_aligned(asize_t sz, int modulo,
+                                         caml_stat_block *b)
+{
+  void *result = caml_stat_alloc_aligned_noexc(sz, modulo, b);
   /* malloc() may return NULL if size is 0 */
-  if (result == NULL && sz != 0) caml_raise_out_of_memory ();
+  if ((result == NULL) && (sz != 0))
+    caml_raise_out_of_memory();
+  return result;
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_alloc_noexc(asize_t sz)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return malloc(sz);
+
+  struct pool_block *pb = malloc(sz + SIZEOF_POOL_BLOCK);
+  if (pb == NULL) return NULL;
 #ifdef DEBUG
-  memset (result, Debug_uninit_stat, sz);
+  memset(&(pb->data), Debug_uninit_stat, sz);
+  pb->magic = Debug_pool_magic;
 #endif
+
+  /* Linking the block into the ring */
+  pb->next = pool->next;
+  pb->prev = pool;
+  pool->next->prev = pb;
+  pool->next = pb;
+
+  return &(pb->data);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_alloc(asize_t sz)
+{
+  void *result = caml_stat_alloc_noexc(sz);
+  /* malloc() may return NULL if size is 0 */
+  if ((result == NULL) && (sz != 0))
+    caml_raise_out_of_memory();
   return result;
 }
 
-CAMLexport void caml_stat_free (void * blk)
+CAMLexport void caml_stat_free(caml_stat_block b)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return free(b);
+
+  struct pool_block *pb = get_pool_block(b);
+  if (pb == NULL) return;
+
+  /* Unlinking the block from the ring */
+  pb->prev->next = pb->next;
+  pb->next->prev = pb->prev;
+
+  free(pb);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_resize_noexc(caml_stat_block b, asize_t sz)
+{
+  // Backward compatibility mode
+  if (pool == NULL) return realloc(b, sz);
+
+  struct pool_block *pb = get_pool_block(b);
+  struct pool_block *pb_new = realloc(pb, sz + SIZEOF_POOL_BLOCK);
+  if (pb_new == NULL) return NULL;
+
+  /* Relinking the new block into the ring in place of the old one */
+  pb_new->prev->next = pb_new;
+  pb_new->next->prev = pb_new;
+
+  return &(pb_new->data);
+}
+
+/* [sz] is a number of bytes */
+CAMLexport caml_stat_block caml_stat_resize(caml_stat_block b, asize_t sz)
 {
-  free (blk);
+  void *result = caml_stat_resize_noexc(b, sz);
+  if (result == NULL)
+    caml_raise_out_of_memory();
+  return result;
 }
 
 /* [sz] is a number of bytes */
-CAMLexport void * caml_stat_resize (void * blk, asize_t sz)
+CAMLexport caml_stat_block caml_stat_calloc_noexc(asize_t num, asize_t sz)
 {
-  void * result = realloc (blk, sz);
+  sz *= num;
```

Comment:
Checking for overflow in a portable way is non-trivial in C. The popular CERT C Coding Standard mentions a [simple implementation](https://www.securecoding.cert.org/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow) of the integer multiplication overflow detection algorithm that I've implemented in OCaml [recently](https://github.com/ocaml-batteries-team/batteries-included/pull/583). The algorithm is not terribly efficient, but should be fine for a start. If it appears that the simple algorithm causes a performance regression, it could be replaced in the future with a more complicated one.

Shall I put the overflow detection functions into `misc`?